storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java

```diff
@@ -489,14 +498,28 @@ public boolean isInitialized() {
         return initialized.get();
     }
 
-    private void ensureInitializedAndNotClosed() {
+    private void ensureInitializedAndNotClosed() throws RemoteStorageException {
         if (initializationFailed) {
             // If initialization is failed, shutdown the broker.
             throw new FatalExitError();
         }
-        if (closing.get() || !initialized.get()) {
-            throw new IllegalStateException("This instance is in invalid state, initialized: " + initialized +
-                                                    " close: " + closing);
+
+        if (!initialized.get()) {
+            log.info("Waiting for TopicBasedRemoteLogMetadataManager to initialize");
+            try {
+                if (!initializeLatch.await(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
```

Comment:
Right, that was the reason for failing the requests immediately. We do not want to hold on to those threads until the RLMM initialises.

Reply:
Thanks for the response. While I understand the problem regarding holding up the threads because the number of calling threads may be limited unless I am missing something in practice this shouldn't be a big concern because in most cases you would either get a single block or non at all (i.e. if the methods are called after `configure` is initialized then `initializeLatch.await` won't block anything and there won't be any context switching either).

If we also think about this problem in another context, i.e. from the perspective of Kafka users/creators of the `RemoteLogMetadataManager` with custom plugins this behaviour is not exactly satisfactory if we for example take into account how other plugin based ecosystems like connect work. Such behaviour is going to create lots of workarounds/hacks (even in our case of using Tiered Storage, with this change `TieredStorageTestHarness` fails to run properly when running in different Kafka versions because unless you put random sleeps because it has no idea when `RemoteLogMetadataManager` has been initialized since all fields/data for initialization is package private). So how long something takes to initialize is a block box aside from TopicBasedRemoteLogMetadataManagerHarness which can peek into package private.

If blocking the thread is such a blocker (pun intended!) would it be more amicable to adjust the interface so its possible for plugin authors (and others) to figure out when any `RemoteLogMetadataManager` has been initialized? If not blocking the thread is so critical then to me its appropriate then to expose the intialization so users can actually handle this case comprehensively.

Reply:
Another alternative option would be to make the `configure` method blocking, this means we don't have to deal with the issue of the other methods with the `RemoteLogMetadataManager` blocking scarce caller threads however I suspect that there is an underlying reason why `configure` is not blocking in the first place?